include/ghex/transport_layer/ucx/request.hpp

```diff
@@ -272,38 +279,69 @@ namespace gridtools{
+                                m_completed.reset();
+                                return false;
                             }
+
+                            if (m_req->m_kind == request_kind::send) return false;
+                            
+                            // set flag to cancelled
+                            *m_completed = 3;
+                                
+                            auto ucx_ptr = m_req->m_ucx_ptr;
+                            auto worker = m_req->m_worker->get();
+
+                            // set to zero here????
+                            // if we assume that the callback is always called, we
+                            // can handle this in the callback body- otherwise needs
+                            // to be done here:
+                            //request_init(ucx_ptr);
+                            m_req->m_kind = request_kind::none;
+                            ucp_request_cancel(worker, ucx_ptr);
                             m_req = nullptr;
                             m_completed.reset();
                             return true;
-                        });
+                        }
+
+                        //// TODO at this time, send requests cannot be canceled
+                        //// in UCX (1.7.0rc1)
+                        //// https://github.com/openucx/ucx/issues/1162
+                        ////
+                        //// TODO the below is only correct for recv requests,
+                        //// or for send requests under the assumption that
+                        //// the requests cannot be moved between threads.
+                        ////
+                        //// For the send worker we do not use locks, hence
+                        //// if request is canceled on another thread, it might
+                        //// clash with another send being submitted by the owner
+                        //// of ucp_worker
+
+                        //if (m_req->m_kind == request_kind::send) return false;
+
+                        //return m_req->m_worker->m_thread_primitives->critical([this]() {
+                        //    if (!(*m_completed)) {
+                        //        auto ucx_ptr = m_req->m_ucx_ptr;
+                        //        auto worker = m_req->m_worker->get();
+
+                        //        // set to zero here????
+                        //        // if we assume that the callback is always called, we
+                        //        // can handle this in the callback body- otherwise needs
+                        //        // to be done here:
+                        //        //request_init(ucx_ptr);
+                        //        ucp_request_cancel(worker, ucx_ptr);
+                        //    }
+                        //    m_req = nullptr;
+                        //    m_completed.reset();
+                        //    return true;
+                        //});
```

Comment:
Is this part relevant?

Reply:
at least the comment is - we can't cancel send requests